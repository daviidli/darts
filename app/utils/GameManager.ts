/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
