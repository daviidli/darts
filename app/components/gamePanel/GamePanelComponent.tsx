import React, { useEffect } from 'react';
import GameManager from '../../utils/GameManager';
import Panel from '../panel/Panel';
import Winner from '../../containers/Winner';
import ClearDarts from '../clearDarts/ClearDarts';

export type Props = {
	winner: number;
	waiting: boolean;
	gameManager: GameManager;
};

const GamePanelComponent = (props: Props) => {
	const { winner, waiting, gameManager } = props;

	useEffect(() => {
		gameManager.start();
	}, [gameManager]);

	if (winner !== -1) {
		return <Winner />;
	}

	if (waiting) {
		return <ClearDarts gameManager={gameManager} waiting={waiting} />;
	}

	if (!gameManager.getRunning()) {
		return <div>starting up</div>;
	}

	return <Panel gameManager={gameManager} />;
};

export default GamePanelComponent;
