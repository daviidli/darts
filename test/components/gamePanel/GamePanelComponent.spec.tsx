import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import GamePanelComponent, {
	Props
} from '../../../app/components/gamePanel/GamePanelComponent';
import GameManager from '../../../app/utils/GameManager';

Enzyme.configure({ adapter: new Adapter() });

const initialProps: Props = {
	winner: -1,
	waiting: false,
	gameManager: { getRunning: () => true } as GameManager
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<GamePanelComponent {...props} />);
	return { wrapper };
};

const setupMount = (props: Props = initialProps) => {
	const wrapper = mount(<GamePanelComponent {...props} />);
	return { wrapper };
};

describe('GamePanel component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot when waiting = true', () => {
		const props = { ...initialProps };
		props.waiting = true;
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot when winner != -1', () => {
		const props = { ...initialProps };
		props.winner = 0;
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot when gameManager is not running', () => {
		const props = { ...initialProps };
		props.gameManager = { getRunning: () => false } as GameManager;
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should call gameManager.start() on load', () => {
		const startMock = spy();
		const props = { ...initialProps };
		props.gameManager = {
			getRunning: () => false,
			start: startMock
		} as any;
		expect(startMock.notCalled).toBeTruthy();
		setupMount(props);
		expect(startMock.calledOnce).toBeTruthy();
	});
});
