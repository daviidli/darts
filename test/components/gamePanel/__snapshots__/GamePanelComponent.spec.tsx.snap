// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePanel component should match snapshot 1`] = `
<Panel
  gameManager={
    Object {
      "getRunning": [Function],
    }
  }
/>
`;

exports[`GamePanel component should match snapshot when gameManager is not running 1`] = `
<div>
  starting up
</div>
`;

exports[`GamePanel component should match snapshot when waiting = true 1`] = `
<ClearDarts
  gameManager={
    Object {
      "getRunning": [Function],
    }
  }
  waiting={true}
/>
`;

exports[`GamePanel component should match snapshot when winner != -1 1`] = `<Connect(WinnerComponent) />`;
